File: python/interval_list_intersections.py
class Solution:
    def intervalIntersection(
        self, firstlist: list[list[int]], secondlist: list[list[int]]
    ) -> list[list[int]]:
        """time complexity: O(n)"""
        i: int = 0
        j: int = 0
        result: list[int] = []
        while i < len(firstlist) and j < len(secondlist):
            a_s, a_e = firstlist[i]
            b_s, b_e = secondlist[j]
            if a_s <= b_e and a_e >= b_s:
                result.append([max(a_s, b_s), min(a_e, b_e)])
            if a_e <= b_e:
                i += 1
            else:
                j += 1
        return result


def test_intervalIntersection_case_1():
    # arrange
    firstList: list[list[int]] = [[0, 2], [5, 10], [13, 23], [24, 25]]
    secondList: list[list[int]] = [[1, 5], [8, 12], [15, 24], [25, 26]]
    expected: list[int] = [[1, 2], [5, 5], [8, 10], [15, 23], [24, 24], [25, 25]]

    # act
    actual = Solution().intervalIntersection(firstList, secondList)

    # assert
    assert expected == actual


def test_intervalIntersection_case_2():
    # arrange
    firstList: list[list[int]] = [[1, 3], [5, 9]]
    secondList: list[list[int]] = []
    expected: list[int] = []

    # act
    actual = Solution().intervalIntersection(firstList, secondList)

    # assert
    assert expected == actual

File: rust/src/code/interval_list_intersections.rs
struct Solution;

impl Solution {
    // time complexity: O(n)
    pub fn interval_intersection(
        first_list: Vec<Vec<i32>>,
        second_list: Vec<Vec<i32>>,
    ) -> Vec<Vec<i32>> {
        let mut result: Vec<Vec<i32>> = Vec::new();
        let mut i: usize = 0;
        let mut j: usize = 0;
        while i < first_list.len() && j < second_list.len() {
            if first_list[i][0] <= second_list[j][1] && first_list[i][1] >= second_list[j][0] {
                result.push(vec![
                    std::cmp::max(first_list[i][0], second_list[j][0]),
                    std::cmp::min(first_list[i][1], second_list[j][1]),
                ]);
            }
            if first_list[i][1] <= second_list[j][1] {
                i += 1
            } else {
                j += 1
            }
        }
        result
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test_interval_intersection_case_1() {
        // arrange
        let firstList: Vec<Vec<i32>> = vec![vec![0, 2], vec![5, 10], vec![13, 23], vec![24, 25]];
        let secondList: Vec<Vec<i32>> = vec![vec![1, 5], vec![8, 12], vec![15, 24], vec![25, 26]];
        let expected: Vec<Vec<i32>> = vec![
            vec![1, 2],
            vec![5, 5],
            vec![8, 10],
            vec![15, 23],
            vec![24, 24],
            vec![25, 25],
        ];

        // act
        let actual = Solution::interval_intersection(firstList, secondList);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_interval_intersection_case_2() {
        // arrange
        let firstList: Vec<Vec<i32>> = vec![vec![0, 2], vec![5, 10], vec![13, 23], vec![24, 25]];
        let secondList: Vec<Vec<i32>> = vec![];
        let expected: Vec<Vec<i32>> = vec![];

        // act
        let actual = Solution::interval_intersection(firstList, secondList);

        // assert
        assert_eq!(expected, actual);
    }
}
